overlord/snapstate/snapstate.go

```diff
@@ -1358,6 +1358,90 @@ func installWithDeviceContext(st *state.State, name string, opts *RevisionOption
 	return doInstall(st, &snapst, snapsup, 0, fromChange, nil)
 }
 
+func Download(ctx context.Context, st *state.State, name string, blobDirectory string, opts *RevisionOptions, userID int, flags Flags, deviceCtx DeviceContext) (*state.TaskSet, error) {
+	if opts == nil {
+		opts = &RevisionOptions{}
+	}
+
+	if opts.CohortKey != "" && !opts.Revision.Unset() {
+		return nil, errors.New("cannot specify revision and cohort")
+	}
+
+	if opts.Channel == "" {
+		opts.Channel = "stable"
+	}
+
+	var snapst SnapState
+	err := Get(st, name, &snapst)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return nil, err
+	}
+
+	if err := snap.ValidateInstanceName(name); err != nil {
+		return nil, fmt.Errorf("invalid instance name: %v", err)
+	}
+
+	sar, err := downloadInfo(ctx, st, name, opts, userID, deviceCtx)
+	if err != nil {
+		return nil, err
+	}
+
+	info := sar.Info
+
+	if blobDirectory == "" {
+		blobDirectory = dirs.SnapBlobDir
+	}
+
+	// if we are going to use the default download dir, and the same snap
+	// revision is already installed, then we should not overwrite the snap that
+	// is already in the dir.
+	if blobDirectory == dirs.SnapBlobDir && info.Revision == snapst.Current {
+		return nil, &snap.AlreadyInstalledError{Snap: name}
+	}
+
+	if flags.RequireTypeBase && info.Type() != snap.TypeBase && info.Type() != snap.TypeOS {
+		return nil, fmt.Errorf("unexpected snap type %q, instead of 'base'", info.Type())
+	}
+
+	if err := checkDiskSpaceDownload([]minimalInstallInfo{installSnapInfo{info}}, blobDirectory); err != nil {
```

Comment:
Checking the disk space in the `Install` functions is feature flagged. Should this be feature flagged? Or should be even do it at all?

Reply:
Also, there is already a `checkDiskSpace` that works for installing. I'm not sure if these should be combined or not. It does operate on different changes kinds (install, refresh), but it also accounts for prerequisites of the snap that is being installed/refreshed (which we don't want in this case, I think).

Reply:
this seems ok/reasonable, if it's just checking that the file will fit on the target disk?

Reply:
Yes, that is all it is doing.